org.spoofax.terms/src/org/spoofax/terms/StrategoArrayList.java

```diff
@@ -1,289 +1,96 @@
 package org.spoofax.terms;
 
+import org.spoofax.interpreter.terms.IStrategoArrayList;
 import org.spoofax.interpreter.terms.IStrategoList;
 import org.spoofax.interpreter.terms.IStrategoTerm;
 import org.spoofax.interpreter.terms.ITermPrinter;
 import org.spoofax.terms.attachments.ITermAttachment;
-import org.spoofax.terms.attachments.TermAttachmentType;
-import org.spoofax.terms.util.TermUtils;
-import java.io.IOException;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.List;
-import java.util.NoSuchElementException;
-import java.util.RandomAccess;
-
-import javax.annotation.Nullable;
 
-import static org.spoofax.terms.AbstractTermFactory.EMPTY_TERM_ARRAY;
+import java.io.IOException;
 
-public class StrategoArrayList extends StrategoTerm implements IStrategoList, RandomAccess {
+/**
+ * @see IStrategoArrayList for most of the shared implementation for this class and {@link StrategoArrayListTail}
+ */
+public class StrategoArrayList extends StrategoTerm implements IStrategoArrayList {
     private static final long serialVersionUID = -1746012089187246512L;
 
-    final IStrategoTerm[] terms;
+    final IStrategoTerm[] backingArray;
     private final int offset;
     final int endOffset;
-    final ITermAttachment[] tailAttachments;
-
-    public StrategoArrayList(IStrategoTerm... terms) {
-        this(terms, null);
-    }
+    ITermAttachment[] tailAttachments;
 
-    public StrategoArrayList(IStrategoTerm[] terms, IStrategoList annotations) {
-        this(terms, annotations, 0, terms.length, new ITermAttachment[terms.length]);
+    public StrategoArrayList(IStrategoTerm... backingArray) {
+        this(backingArray, null);
     }
```

Comment:
Could the constructors be `package private`? These are implementation details and should never be called directly. Also, this constructor is never called as a `varargs` constructor, so they should have the type `StrategoArrayList(IStrategoTerm[] backingArray)`.